lib/libvfio-user.c

```diff
@@ -536,15 +541,19 @@ handle_dma_map_or_unmap(vfu_ctx_t *vfu_ctx, uint32_t size, bool map,
     }
 
     nr_dma_regions = (int)(size / sizeof(struct vfio_user_dma_region));
+    if (nr_dma_regions == 0) {
+        return 0;
+    }
 
     for (i = 0, fdi = 0; i < nr_dma_regions; i++) {
         if (map) {
             int fd = -1;
             if (dma_regions[i].flags == VFIO_USER_F_DMA_REGION_MAPPABLE) {
                 if (fdi == nr_fds) {
-                    return -EINVAL;
+                    ret = -EINVAL;
+                    break;
                 }
-                fd = fds[fdi++];
+                fd = fds[fdi];
```

Comment:
this I really don't like though! The whole point of consume_fd() was that it happens at the point we deference the fd array.

If dma_controller_add_region() fails, then we should explicitly close(fd) that we just took ownership of. That's much more obvious pattern than what's here now.

Reply:
right, will fix